### PR TITLE
Issue 242: Move dialogue feedback type selection widget to interstitial before tdfs with dialogue enabled and allow feedback type choice enabled

### PR DESCRIPTION
**Pull request by JRustyHaner**
#242 fixes issue with initial pull request, dynamically reading if a tdf allows for feedback type

### CODE REVIEW

File: mofacts/client/views/experiment/card.js [hunk 1]
```diff
@@ -372,7 +372,16 @@ Template.card.rendered = async function() {
   }
 
   // Check if TDF allows for dialogue feedback preferences, if so, route to dialogue feedback widget
-  if(Session.get('allowFeedbackTypeSelect') && !Session.get('feedbackParamsSet')) {
+  let allowFeedbackTypeSelect = false;
+  let feedbackUnitNumLookup = 0;
+
+  if(typeof Session.get('currentUnitNumber') !== 'undefined'){
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:378]: currentUnitNumber is only going to be set somewhat into card.js, in resumeComponentState.  At that point the deliveryParams have also been read in and you can use getCurrentDeliveryParams.  Can you refactor this to happen at the end of resumeFromComponentState by jumping to the feedback page and doing a check similar to the end of Template.card.rendered but with a bit to then go to processUserTimesLog if we jumped out to feedback before doing the syllable cache check and it?

**Review by andrewtackett — CHANGES_REQUESTED** [review]

File: mofacts/client/views/experiment/card.js [hunk 2]
```diff
@@ -2666,12 +2660,31 @@ async function resumeFromComponentState() {
 
   await updateExperimentState(newExperimentState, 'card.resumeFromComponentState');
 
+  //check if user feedback settins are enabled, if so redirect to settings menu
+
+  await getFeebackParameters(curTdfUnit);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2665]: Doesn't need to be async (await/async)

File: mofacts/client/views/experiment/card.js [hunk 3]
```diff
@@ -2666,12 +2660,31 @@ async function resumeFromComponentState() {
...
+
   // Notice that no matter what, we log something about condition data
   // ALSO NOTICE that we'll be calling processUserTimesLog after the server
   // returns and we know we've logged what happened
   checkSyllableCacheForCurrentStimFile(processUserTimesLog);
 }
 
+
+
+async function getFeebackParameters(curTdfUnit){
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2675]: feedback is misspelled

**Comment by JRustyHaner** [comment]
@andrewtackett

File: mofacts/client/views/experiment/card.js [hunk 4]
```diff
@@ -2666,12 +2660,31 @@ async function resumeFromComponentState() {
...
+  if(typeof curTdfUnit.deliveryparams[0].allowFeedbackTypeSelect !== 'undefined'){
+    allowFeedbackTypeSelect = curTdfUnit.deliveryparams[0].allowFeedbackTypeSelect[0];
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2677]: Use getCurrentDeliveryParams()

**Review by andrewtackett — COMMENTED** [review]
How does this selectively jump into processUserTimesLog like I mentioned?

**Comment by JRustyHaner** [comment]
@andrewtackett

File: mofacts/client/views/experiment/card.js [hunk 5]
```diff
@@ -2665,13 +2660,24 @@ async function resumeFromComponentState() {
...
+  //check if user feedback settins are enabled, if so redirect to settings menu
+  await getFeedbackParameters(curTdfUnit);
 
   // Notice that no matter what, we log something about condition data
   // ALSO NOTICE that we'll be calling processUserTimesLog after the server
   // returns and we know we've logged what happened
   checkSyllableCacheForCurrentStimFile(processUserTimesLog);
 }
 
+async function getFeedbackParameters(curTdfUnit){
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2673]: Remove async here and await at function call, this isn't asynchronous code

**Review by andrewtackett — CHANGES_REQUESTED** [review]
Need to handle popping out to feedback and resuming to past where we pop out to feedback.  You should be able to do this with a session variable check, probably in rendered

**Comment by JRustyHaner** [comment]
@andrewtackett

File: mofacts/client/views/experiment/card.js [hunk 6]
```diff
@@ -2661,20 +2662,21 @@ async function resumeFromComponentState() {
...
-  await getFeedbackParameters(curTdfUnit);
+
 
   // Notice that no matter what, we log something about condition data
   // ALSO NOTICE that we'll be calling processUserTimesLog after the server
   // returns and we know we've logged what happened
   checkSyllableCacheForCurrentStimFile(processUserTimesLog);
+
+  //check if user feedback settins are enabled, if so redirect to settings menu
+  getFeedbackParameters(curTdfUnit);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2673]: @JRustyHaner Did you test this?

Comment (JRustyHaner) [reply mofacts/client/views/experiment/card.js:2673]: Tested on testtdf and a tdf that did not have the allowFeedbackTypeSelect

Comment (andrewtackett) [reply mofacts/client/views/experiment/card.js:2673]: You don't have branch selection here. You need to selectively go down one path or the other

File: mofacts/client/views/experiment/card.html [hunk 7]
```diff
@@ -1,6 +1,13 @@
...
     <div class="container" id="cardContainer">
-
+    
+    {{#if displayFeedback}}
+        {{>profileDialogueToggles}}
+        <div>
+            <br />
+            <button id='confirmFeedbackSelection' class='btn'>Confirm feedback settings</button>
+        </div>
+    {{/if}}
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.html:10]: I like to make it very explictly clear when two paths diverge.  Can you put the normal card template html into an else condition here so it's very obvious that these are supposed to be separate portions of the card loop?

File: mofacts/client/views/experiment/card.js [hunk 8]
```diff
@@ -442,7 +436,11 @@ Template.card.events({
   'keypress #userForceCorrect': function(e) {
     handleUserForceCorrectInput(e, 'keypress');
   },
-
+  'click #confirmFeedbackSelection': function() {
+    Session.set('displayFeedback', false);
+    checkSyllableCacheForCurrentStimFile(processUserTimesLog);  
+
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:442]: Nitpick, extra line of whitespace

File: mofacts/client/views/experiment/card.js [hunk 9]
```diff
@@ -2666,12 +2664,30 @@ async function resumeFromComponentState() {
 
   await updateExperimentState(newExperimentState, 'card.resumeFromComponentState');
 
+  getFeebackParameters(curTdfUnit);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2667]: I think when you rolled back some it took out the spelling correction of feedback here?

File: mofacts/client/views/experiment/card.js (hunk 4 above)
Comment (andrewtackett) [reply mofacts/client/views/experiment/card.js:2680]: Looks like the same happened with using getCurrentDeliveryParams here

**Review by andrewtackett — CHANGES_REQUESTED** [review]
A few minor changes and it'll be good to go

**Comment by JRustyHaner** [comment]
@andrewtackett yeah I was working on an old commit, updated

File: mofacts/client/views/experiment/card.html (hunk 7 above)
Comment (andrewtackett) [reply mofacts/client/views/experiment/card.html:9]: ^

File: mofacts/client/views/experiment/card.js [hunk 10]
```diff
@@ -2666,12 +2663,30 @@ async function resumeFromComponentState() {
...
   // ALSO NOTICE that we'll be calling processUserTimesLog after the server
   // returns and we know we've logged what happened
-  checkSyllableCacheForCurrentStimFile(processUserTimesLog);
+  if(!Session.get('displayFeedback')){
+    checkSyllableCacheForCurrentStimFile(processUserTimesLog);
+  }
 }
 
+
+async function getFeebackParameters(){
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:2677]: Still misspelled

**Review by andrewtackett — CHANGES_REQUESTED** [review]

File: mofacts/client/views/experiment/card.html (hunk 7 above)
Comment (andrewtackett) [reply mofacts/client/views/experiment/card.html:9]: ^

**Review by andrewtackett — CHANGES_REQUESTED** [review]

**Comment by JRustyHaner** [comment]
@andrewtackett

**Review by andrewtackett — APPROVED** [review]

**Pull request merged**